File: cfr/abstract_game.py
class AbstractGame:
    def get_initial_state(self):
        """Return the initial state of the game."""
        raise NotImplementedError

    def is_terminal(self, state):
        """Return True if state is terminal."""
        raise NotImplementedError

    def get_utility(self, state):
        """
        Return a dictionary mapping player IDs to their utility
        at the terminal state.
        """
        raise NotImplementedError

    def is_chance_node(self, state):
        """Return True if the state is a chance node (e.g. a random event)."""
        raise NotImplementedError

    def sample_chance_action(self, state):
        """For chance nodes, sample and return an action (or outcome)."""
        raise NotImplementedError

    def get_current_player(self, state):
        """Return the player (or chance) whose turn it is to act."""
        raise NotImplementedError

    def get_information_set(self, state, player):
        """
        Return a key (e.g. a string) that uniquely represents the player’s information set
        in this state.
        """
        raise NotImplementedError

    def get_actions(self, state):
        """
        Return a list of actions available at the given state.
        This allows the set of actions to vary with state.
        """
        raise NotImplementedError

    def apply_action(self, state, action):
        """Return the new state after the given action is applied."""
        raise NotImplementedError

    def get_players(self):
        """Return a list of players in the game."""
        raise NotImplementedError

File: rust/src/abstract_game.rs
// /// A trait representing an abstract game. The type parameter `State` represents
// /// the type used to encode a game state.
// pub trait AbstractGame<State> {
//     fn is_terminal(&self, state: &State) -> bool;
//     fn get_utility(&self, state: &State) -> Vec<f64>;
//     fn is_chance_node(&self, state: &State) -> bool;
//     fn sample_chance_action(&self, state: &State) -> (usize, f64);
//     fn get_child(&self, state: &State, action: usize) -> State;
//     fn get_current_player(&self, state: &State) -> usize;
//     fn get_information_set(&self, state: &State, player: usize) -> usize;
//     fn get_actions(&self, state: &State) -> Vec<usize>;
//     fn get_initial_state(&self) -> State;
//     fn apply_action(&self, state: &State, action: usize) -> State;
// }

/// The abstract game trait.
pub trait AbstractGame<State> {
    fn is_terminal(&self, state: &State) -> bool;
    fn get_utility(&self, state: &State) -> Vec<f64>;
    fn is_chance_node(&self, state: &State) -> bool;
    fn sample_chance_action(&self, state: &State) -> (usize, f64);
    fn get_child(&self, state: &State, action: usize) -> State;
    fn get_current_player(&self, state: &State) -> i32;
    fn get_information_set(&self, state: &State, player: i32) -> usize;
    fn get_actions(&self, state: &State) -> Vec<usize>;
    fn get_initial_state(&self) -> State;
    fn apply_action(&self, state: &State, action: usize) -> State;
}